### ISSUE TITLE
Strange behavior of hipMemcpyAsync on gfx1034

### ISSUE THREAD

**Issue by al42and** [ROCm-Developer-Tools/HIP#3314]
I am experiencing a problem when using `hipMemcpyAsync` with ROCm 5.6.0 on a gfx1034 device (RX 6400; acknowledge that it is not supported on Linux).

The program below allocates two pinned host arrays and a device array. It fills the first host array, copies it asynchronously to the device, then copies back to the second array, synchronizes, then checks for the correctness. In my case, the output is all-zeroes.

What fixes the problem:

- Using non-pinned memory (not practical in real use cases).
- Adding `hipStreamSynchronize` or `hipDeviceSynchronize` between the copies.
- Setting `AMD_SERIALIZE_COPY=2`
- Running on MI50
- Adding `sleep(2);` between the copies

What does **not** fix the problem:

- Setting `AMD_SERIALIZE_COPY=1`
- Adding `sleep(2);` before checking the results

```cpp
#include <hip/hip_runtime.h>
#include <iostream>

const int N = 10;

#define hip_check(ex)                                                                              \
  do {                                                                                             \
    hipError_t status = ex;                                                                        \
    if (status != hipSuccess) {                                                                    \
      fprintf(stderr, "Error: %s at %s:%d\n", #ex, __FILE__, __LINE__);                            \
      const char *msg = hipGetErrorString(status);                                                 \
      fprintf(stderr, "%s\n", msg);                                                                \
    }                                                                                              \
  } while (0)

int main() {
  hipStream_t stream;
  hip_check(hipStreamCreate(&stream));

  // Allocate memory on the host
  int *hostData, *hostData2;
  hip_check(hipHostMalloc(&hostData, N * sizeof(int)));
  hip_check(hipHostMalloc(&hostData2, N * sizeof(int)));
  // Using non-pinned memory avoids the bug:
  // hostData = (int*)malloc(N * sizeof(int));
  // hostData2 = (int*)malloc(N * sizeof(int));

  // Initialize host data
  for (int i = 0; i < N; ++i)
    hostData[i] = i;

  // Allocate buffer on the device
  int *deviceData;
  hip_check(hipMalloc(&deviceData, N * sizeof(int)));

  // Copy data from host to device
  hip_check(hipMemcpyAsync(deviceData, hostData, N * sizeof(int), hipMemcpyHostToDevice, stream));

  // Uncommenting the sync below fixes the problem
  // hipStreamSynchronize(stream);

  // Copy data back from device to host
  hip_check(hipMemcpyAsync(hostData2, deviceData, N * sizeof(int), hipMemcpyDeviceToHost, stream));

  hip_check(hipDeviceSynchronize());

  // Compare results
  bool resultsMatch = true;
  for (int i = 0; i < N; ++i) {
    std::cout << hostData[i] << " - " << hostData2[i] << std::endl;
    if (hostData2[i] != i) {
      resultsMatch = false;
    }
  }

  if (resultsMatch) {
    std::cout << "PASS" << std::endl;
  } else {
    std::cout << "FAIL" << std::endl;
  }

  // Clean up
  hip_check(hipFree(hostData));
  hip_check(hipFree(hostData2));
  hip_check(hipFree(deviceData));

  return !resultsMatch;
}
```

```console
$ hipcc --version
HIP version: 5.6.31061-8c743ae5d
AMD clang version 16.0.0 (https://github.com/RadeonOpenCompute/llvm-project roc-5.6.0 23243 be997b2f3651a41597d7a41441fff8ade4ac59ac)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/rocm//llvm/bin

$ hipconfig
HIP version  : 5.6.31061-8c743ae5d

== hipconfig
HIP_PATH     : /opt/rocm-5.6.0
ROCM_PATH    : /opt/rocm/
HIP_COMPILER : clang
HIP_PLATFORM : amd
HIP_RUNTIME  : rocclr
CPP_CONFIG   :  -D__HIP_PLATFORM_HCC__= -D__HIP_PLATFORM_AMD__= -I/opt/rocm-5.6.0/include -I/opt/rocm//llvm/bin/../lib/clang/16.0.0 

== hip-clang
HIP_CLANG_PATH   : /opt/rocm//llvm/bin
AMD clang version 16.0.0 (https://github.com/RadeonOpenCompute/llvm-project roc-5.6.0 23243 be997b2f3651a41597d7a41441fff8ade4ac59ac)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/rocm//llvm/bin
AMD LLVM version 16.0.0git
  Optimized build.
  Default target: x86_64-unknown-linux-gnu
  Host CPU: alderlake

  Registered Targets:
    amdgcn - AMD GCN GPUs
    r600   - AMD GPUs HD2XXX-HD6XXX
    x86    - 32-bit X86: Pentium-Pro and above
    x86-64 - 64-bit X86: EM64T and AMD64
hip-clang-cxxflags :  -isystem "/opt/rocm-5.6.0/include" -O3
hip-clang-ldflags  :  -O3 --hip-link --rtlib=compiler-rt -unwindlib=libgcc

=== Environment Variables
PATH=/home/aland/.cargo/bin:/home/aland/local/bin:/usr/local/cuda/bin:/opt/rocm//bin:/home/aland/.cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
LD_LIBRARY_PATH=/home/aland/local/lib64:/home/aland/local/lib:/usr/local/cuda/lib64:/opt/rocm//lib
CUDA_HOME=/usr/local/cuda

== Linux Kernel
Hostname     : tcbl16.scilifelab.se
Linux tcbl16.scilifelab.se 6.2.0-26-generic #26~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Jul 13 16:27:29 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.3 LTS
Release:        22.04
Codename:       jammy

$ hipcc hipcopy.cpp -o hipcopy && ./hipcopy
0 - 0
1 - 0
2 - 0
3 - 0
4 - 0
5 - 0
6 - 0
7 - 0
8 - 0
9 - 0
FAIL
```

Output of `rocprof --hip-trace --hsa-trace ./hipcopy`: 
[results.json.gz](https://github.com/ROCm-Developer-Tools/HIP/files/12409631/results.json.gz)

**Comment by cjatin** [comment]
thats weird.

Can you provide me with few more things. 

1) complete compile command. run hipcc with `HIPCC_VERBOSE=1 hipcc hipcopy.cpp -o hipcopy`

This should show the clang call.

2) logs after running the exe with `AMD_LOG_LEVEL=7 ./hipcopy`

**Comment by al42and** [comment]
```console
$ HIPCC_VERBOSE=1 hipcc hipcopy.cpp -o hipcopy
hipcc-cmd: /opt/rocm//llvm/bin/clang++  -isystem "/opt/rocm-5.6.0/include" --offload-arch=gfx1034 -O3 -mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false  -O3 --hip-link --rtlib=compiler-rt -unwindlib=libgcc  -x hip hipcopy.cpp -o "hipcopy"

$ AMD_LOG_LEVEL=7 ./hipcopy
:3:rocdevice.cpp            :434 : 1316416763543 us: 3327951: [tid:0x7f4a18571100] Initializing HSA stack.
:3:comgrctx.cpp             :33  : 1316416766754 us: 3327951: [tid:0x7f4a18571100] Loading COMGR library.
:3:rocdevice.cpp            :200 : 1316416766807 us: 3327951: [tid:0x7f4a18571100] Numa selects cpu agent[0]=0x16f9d50(fine=0x16f5260,coarse=0x16fa660) for gpu agent=0x16fb230
:3:rocdevice.cpp            :1634: 1316416767012 us: 3327951: [tid:0x7f4a18571100] HMM support: 1, xnack: 0, direct host access: 0

:4:rocdevice.cpp            :2012: 1316416767062 us: 3327951: [tid:0x7f4a18571100] Allocate hsa host memory 0x7f4a1a182000, size 0x38
:4:rocdevice.cpp            :2012: 1316416767239 us: 3327951: [tid:0x7f4a18571100] Allocate hsa host memory 0x7f490da00000, size 0x101000
:4:rocdevice.cpp            :2012: 1316416767455 us: 3327951: [tid:0x7f4a18571100] Allocate hsa host memory 0x7f490d800000, size 0x101000
:4:runtime.cpp              :83  : 1316416767465 us: 3327951: [tid:0x7f4a18571100] init
:3:hip_context.cpp          :48  : 1316416767466 us: 3327951: [tid:0x7f4a18571100] Direct Dispatch: 1
:3:rocdevice.cpp            :2818: 1316416767507 us: 3327951: [tid:0x7f4a18571100] number of allocated hardware queues with low priority: 0, with normal priority: 0, with high priority: 0, maximum per priority is: 4
:3:rocdevice.cpp            :2896: 1316416769612 us: 3327951: [tid:0x7f4a18571100] created hardware queue 0x7f4a1a172000 with size 16384 with priority 1, cooperative: 0
:3:rocdevice.cpp            :2963: 1316416769617 us: 3327951: [tid:0x7f4a18571100] acquireQueue refCount: 0x7f4a1a172000 (1)

:4:rocdevice.cpp            :2012: 1316416769814 us: 3327951: [tid:0x7f4a18571100] Allocate hsa host memory 0x7f490c800000, size 0x100000
:3:devprogram.cpp           :2684: 1316416883589 us: 3327951: [tid:0x7f4a18571100] Using Code Object V5.
:3:devprogram.cpp           :2984: 1316416884203 us: 3327951: [tid:0x7f4a18571100] For Init/Fini: Kernel Name: __amd_rocclr_copyBufferAligned
:3:devprogram.cpp           :2984: 1316416884207 us: 3327951: [tid:0x7f4a18571100] For Init/Fini: Kernel Name: __amd_rocclr_copyBufferRect
:3:devprogram.cpp           :2984: 1316416884208 us: 3327951: [tid:0x7f4a18571100] For Init/Fini: Kernel Name: __amd_rocclr_copyBufferRectAligned
:3:devprogram.cpp           :2984: 1316416884209 us: 3327951: [tid:0x7f4a18571100] For Init/Fini: Kernel Name: __amd_rocclr_streamOpsWait
:3:devprogram.cpp           :2984: 1316416884210 us: 3327951: [tid:0x7f4a18571100] For Init/Fini: Kernel Name: __amd_rocclr_copyImage
:3:devprogram.cpp           :2984: 1316416884211 us: 3327951: [tid:0x7f4a18571100] For Init/Fini: Kernel Name: __amd_rocclr_fillBufferAligned
:3:devprogram.cpp           :2984: 1316416884212 us: 3327951: [tid:0x7f4a18571100] For Init/Fini: Kernel Name: __amd_rocclr_copyImage1DA
:3:devprogram.cpp           :2984: 1316416884213 us: 3327951: [tid:0x7f4a18571100] For Init/Fini: Kernel Name: __amd_rocclr_copyImageToBuffer
:3:devprogram.cpp           :2984: 1316416884214 us: 3327951: [tid:0x7f4a18571100] For Init/Fini: Kernel Name: __amd_rocclr_initHeap
:3:devprogram.cpp           :2984: 1316416884215 us: 3327951: [tid:0x7f4a18571100] For Init/Fini: Kernel Name: __amd_rocclr_streamOpsWrite
:3:devprogram.cpp           :2984: 1316416884216 us: 3327951: [tid:0x7f4a18571100] For Init/Fini: Kernel Name: __amd_rocclr_copyBuffer
:3:devprogram.cpp           :2984: 1316416884217 us: 3327951: [tid:0x7f4a18571100] For Init/Fini: Kernel Name: __amd_rocclr_fillBufferAligned2D
:3:devprogram.cpp           :2984: 1316416884218 us: 3327951: [tid:0x7f4a18571100] For Init/Fini: Kernel Name: __amd_rocclr_copyBufferToImage
:3:devprogram.cpp           :2984: 1316416884219 us: 3327951: [tid:0x7f4a18571100] For Init/Fini: Kernel Name: __amd_rocclr_fillImage
:3:hip_stream.cpp           :370 : 1316416884282 us: 3327951: [tid:0x7f4a18571100] hipStreamCreate: Returned hipSuccess : stream:0x161b0a0
:3:hip_memory.cpp           :568 : 1316416884288 us: 3327951: [tid:0x7f4a18571100]  hipHostMalloc ( 0x7ffdef704bf8, 40, 0 ) 
:4:rocdevice.cpp            :2012: 1316416884309 us: 3327951: [tid:0x7f4a18571100] Allocate hsa host memory 0x7f4a18540000, size 0x28
:3:hip_memory.cpp           :610 : 1316416884317 us: 3327951: [tid:0x7f4a18571100] hipHostMalloc: Returned hipSuccess : 0x7f4a18540000: duration: 29 us
:3:hip_memory.cpp           :568 : 1316416884318 us: 3327951: [tid:0x7f4a18571100]  hipHostMalloc ( 0x7ffdef704bf0, 40, 0 ) 
:4:rocdevice.cpp            :2012: 1316416884331 us: 3327951: [tid:0x7f4a18571100] Allocate hsa host memory 0x7f4a1853e000, size 0x28
:3:hip_memory.cpp           :610 : 1316416884336 us: 3327951: [tid:0x7f4a18571100] hipHostMalloc: Returned hipSuccess : 0x7f4a1853e000: duration: 18 us
:3:hip_memory.cpp           :562 : 1316416891013 us: 3327951: [tid:0x7f4a18571100]  hipMalloc ( 0x7ffdef704bc0, 40 ) 
:4:rocdevice.cpp            :2141: 1316416891063 us: 3327951: [tid:0x7f4a18571100] Allocate hsa device memory 0x7f490c400000, size 0x28
:3:rocdevice.cpp            :2180: 1316416891064 us: 3327951: [tid:0x7f4a18571100] device=0x1718a20, freeMem_ = 0xfeffffd8
:3:hip_memory.cpp           :564 : 1316416891066 us: 3327951: [tid:0x7f4a18571100] hipMalloc: Returned hipSuccess : 0x7f490c400000: duration: 53 us
:3:hip_memory.cpp           :1423: 1316416891262 us: 3327951: [tid:0x7f4a18571100]  hipMemcpyAsync ( 0x7f490c400000, 0x7f4a18540000, 40, hipMemcpyHostToDevice, stream:0x161b0a0 ) 
:4:command.cpp              :349 : 1316416891268 us: 3327951: [tid:0x7f4a18571100] Command (CopyDeviceToDevice) enqueued: 0x1d3e830
:4:rocblit.cpp              :697 : 1316416892005 us: 3327951: [tid:0x7f4a18571100] Query copy engine status 0, freemask 3
:4:rocblit.cpp              :707 : 1316416892008 us: 3327951: [tid:0x7f4a18571100] HSA Async Copy on copy_engine=1, dst=0x7f490c400000, src=0x7f4a18540000, size=40, wait_event=0x0, completion_signal=0x7f4a1a193800
:3:hip_memory.cpp           :1424: 1316416892014 us: 3327951: [tid:0x7f4a18571100] hipMemcpyAsync: Returned hipSuccess : : duration: 752 us
:3:hip_memory.cpp           :1423: 1316416892017 us: 3327951: [tid:0x7f4a18571100]  hipMemcpyAsync ( 0x7f4a1853e000, 0x7f490c400000, 40, hipMemcpyDeviceToHost, stream:0x161b0a0 ) 
:4:command.cpp              :349 : 1316416892019 us: 3327951: [tid:0x7f4a18571100] Command (CopyDeviceToDevice) enqueued: 0x1758020
:3:rocvirtual.hpp           :66  : 1316416892021 us: 3327951: [tid:0x7f4a18571100] Host active wait for Signal = (0x7f4a1a193800) for 10000 ns
:4:rocblit.cpp              :697 : 1316416892033 us: 3327951: [tid:0x7f4a18571100] Query copy engine status 0, freemask 2
:4:rocblit.cpp              :707 : 1316416892035 us: 3327951: [tid:0x7f4a18571100] HSA Async Copy on copy_engine=2, dst=0x7f4a1853e000, src=0x7f490c400000, size=40, wait_event=0x7f4a1a193800, completion_signal=0x7f4a1a193780
:1:rocblit.cpp              :736 : 1316416892036 us: 3327951: [tid:0x7f4a18571100] HSA copy from host to device failed with code 4097
:1:rocvirtual.cpp           :1856: 1316416892037 us: 3327951: [tid:0x7f4a18571100] submitCopyMemory failed!
:4:command.cpp              :179 : 1316416892038 us: 3327951: [tid:0x7f4a18571100] Command 0x1758020 complete
:3:hip_memory.cpp           :1424: 1316416892039 us: 3327951: [tid:0x7f4a18571100] hipMemcpyAsync: Returned hipSuccess : : duration: 22 us
:3:hip_device_runtime.cpp   :517 : 1316416892041 us: 3327951: [tid:0x7f4a18571100]  hipDeviceSynchronize (  ) 
:3:rocdevice.cpp            :2737: 1316416892043 us: 3327951: [tid:0x7f4a18571100] No HW event
:4:commandqueue.cpp         :138 : 1316416892044 us: 3327951: [tid:0x7f4a18571100] HW Event not ready, awaiting completion instead
:4:commandqueue.cpp         :151 : 1316416892045 us: 3327951: [tid:0x7f4a18571100] All commands finished
:3:hip_device_runtime.cpp   :519 : 1316416892070 us: 3327951: [tid:0x7f4a18571100] hipDeviceSynchronize: Returned hipSuccess : 
0 - 0
1 - 0
2 - 0
3 - 0
4 - 0
5 - 0
6 - 0
7 - 0
8 - 0
9 - 0
FAIL
:3:hip_memory.cpp           :614 : 1316418892276 us: 3327951: [tid:0x7f4a18571100]  hipFree ( 0x7f4a18540000 ) 
:3:hip_memory.cpp           :616 : 1316418892287 us: 3327951: [tid:0x7f4a18571100] hipFree: Returned hipSuccess : 
:3:hip_memory.cpp           :614 : 1316418892288 us: 3327951: [tid:0x7f4a18571100]  hipFree ( 0x7f4a1853e000 ) 
:3:hip_memory.cpp           :616 : 1316418892290 us: 3327951: [tid:0x7f4a18571100] hipFree: Returned hipSuccess : 
:3:hip_memory.cpp           :614 : 1316418892291 us: 3327951: [tid:0x7f4a18571100]  hipFree ( 0x7f490c400000 ) 
:3:hip_memory.cpp           :616 : 1316418892292 us: 3327951: [tid:0x7f4a18571100] hipFree: Returned hipSuccess : 
```

Interesting, so the second memcpy fails for some reason (presumably, HSA_STATUS_ERROR_INVALID_ARGUMENT), but the error is not propagated :unamused:.

Adding `sleep(2)`  between the two memcpy calls fixes the issue. Is there some problem with running two memory operations simultaneously?

Note: This reduced example always fails for me. But originally I encountered the problem in a larger application, where it was stochastic (frequent, but not always).

**Comment by al42and** [comment]
Upd: If I add `sleep` or `hipStreamSynchronize` between the two memcpy's, then the second is executed on `copy_engine=1` and no dependencies (and succeeds):


```
:3:hip_memory.cpp           :1423: 1319191147091 us: 3332255: [tid:0x7f39dd65d100]  hipMemcpyAsync ( 0x7f39dba40000, 0x7f38cba00000, 40, hipMemcpyDeviceToHost, stream:0x11c10a0 ) 
:4:command.cpp              :349 : 1319191147104 us: 3332255: [tid:0x7f39dd65d100] Command (CopyDeviceToDevice) enqueued: 0x158bab0
:4:rocblit.cpp              :697 : 1319191147114 us: 3332255: [tid:0x7f39dd65d100] Query copy engine status 0, freemask 3
:4:rocblit.cpp              :707 : 1319191147117 us: 3332255: [tid:0x7f39dd65d100] HSA Async Copy on copy_engine=1, dst=0x7f39dba40000, src=0x7f38cba00000, size=40, wait_event=0x0, completion_signal=0x7f39dd694780
:3:hip_memory.cpp           :1424: 1319191147122 us: 3332255: [tid:0x7f39dd65d100] hipMemcpyAsync: Returned hipSuccess : : duration: 31 us
```

**Comment by al42and** [comment]
Looking at [the code](https://github.com/RadeonOpenCompute/ROCR-Runtime/blob/e0fadddb3175cb95ce9e9af2ebd2a205045e854e/src/core/runtime/amd_gpu_agent.cpp#L809-L811), I'm surprised even `copy_engine=1` works, because `engine_offset == ffs(engine_id) == 1`, and

```cpp
  if (engine_offset > properties_.NumSdmaEngines + properties_.NumSdmaXgmiEngines) {
    return HSA_STATUS_ERROR_INVALID_ARGUMENT;
  }
```

```console
$ grep num_sdma_ /sys/devices/virtual/kfd/kfd/topology/nodes/?/properties
/sys/devices/virtual/kfd/kfd/topology/nodes/0/properties:num_sdma_engines 0
/sys/devices/virtual/kfd/kfd/topology/nodes/0/properties:num_sdma_xgmi_engines 0
/sys/devices/virtual/kfd/kfd/topology/nodes/0/properties:num_sdma_queues_per_engine 0
/sys/devices/virtual/kfd/kfd/topology/nodes/1/properties:num_sdma_engines 1
/sys/devices/virtual/kfd/kfd/topology/nodes/1/properties:num_sdma_xgmi_engines 0
/sys/devices/virtual/kfd/kfd/topology/nodes/1/properties:num_sdma_queues_per_engine 8
```

**Comment by cjatin** [comment]
This kind of makes sense.
```
:4:rocblit.cpp              :707 : 1316416892035 us: 3327951: [tid:0x7f4a18571100] HSA Async Copy on copy_engine=2, dst=0x7f4a1853e000, src=0x7f490c400000, size=40, wait_event=0x7f4a1a193800, completion_signal=0x7f4a1a193780
:1:rocblit.cpp              :736 : 1316416892036 us: 3327951: [tid:0x7f4a18571100] HSA copy from host to device failed with code 4097
:1:rocvirtual.cpp           :1856: 1316416892037 us: 3327951: [tid:0x7f4a18571100] submitCopyMemory failed!
```
On your GPU it has 1 sdma engine, where as somehow HIP is trying to run the memcpy on 2nd. Will look into it.

**Comment by cjatin** [comment]
I am really interested in seeing the hsa's SDMA count. Can you share the output of this code:

```cpp
#include <hsa/hsa.h>
#include <hsa/hsa_ext_amd.h>

#include <bitset>
#include <iostream>
#include <map>
#include <string>

#define check_hsa(call_)                                             \
  {                                                                  \
    auto res = (call_);                                              \
    if (res != HSA_STATUS_SUCCESS && res != HSA_STATUS_INFO_BREAK) { \
      std::cout << "Failed in HSA call: " << #call_ << std::endl;    \
    }                                                                \
  }

hsa_status_t iter_agent_callback(hsa_agent_t agent, void* data) {
  auto agent_map = reinterpret_cast<std::map<std::string, hsa_agent_t>*>(data);
  char name[100]{'\0'};
  if (hsa_agent_get_info(agent, HSA_AGENT_INFO_NAME, (void*)name) !=
      HSA_STATUS_SUCCESS) {
    std::cerr << "Failed in hsa_agent_get_info name" << std::endl;
    std::abort();
  }
  std::string s_name{name};
  if (s_name.find("gfx") ==
      std::string::npos) {  // CPU do not have gfx in them generally
    s_name = "cpu";
  }
  (*agent_map)[s_name] = agent;
  return HSA_STATUS_SUCCESS;
}

int main() {
  check_hsa(hsa_init());

  std::map<std::string, hsa_agent_t> agent_map;

  check_hsa(hsa_iterate_agents(iter_agent_callback, (void*)&agent_map));

  std::cout << "Agents found: ";
  for (auto& i : agent_map) {
    std::cout << i.first << ", ";
  }
  std::cout << std::endl;

  std::string s_agent_of_importance = "gfx1034";
  if (agent_map.find(s_agent_of_importance) == agent_map.end()) {
    std::cerr << "gfx arch not found in agent map: " << s_agent_of_importance
              << std::endl;
    std::abort();
  }

  hsa_agent_t h_agent_of_importance = agent_map[s_agent_of_importance];
  hsa_agent_t cpu_agent = agent_map["cpu"];
  std::cout << "Data for CPU to " << s_agent_of_importance << std::endl;

  uint32_t free_engine_mask = 0u;
  check_hsa(hsa_amd_memory_copy_engine_status(h_agent_of_importance, cpu_agent,
                                              &free_engine_mask));
  std::bitset<32> cpu_to_gpu{free_engine_mask};

  free_engine_mask = 0;
  check_hsa(hsa_amd_memory_copy_engine_status(cpu_agent, h_agent_of_importance,
                                              &free_engine_mask));
  std::bitset<32> gpu_to_cpu{free_engine_mask};

  std::cout << "GPU to CPU: " << gpu_to_cpu << std::endl;
  std::cout << "CPU to GPU: " << cpu_to_gpu << std::endl;

  uint32_t count =
      free_engine_mask - (free_engine_mask & (free_engine_mask - 1));
  std::cout << "Count after op: " << count << std::endl;

  check_hsa(hsa_shut_down());
}
```

compile command: hipcc file.cpp -lhsa-runtime64

**Comment by al42and** [comment]
```console
$ hipcc hsa_sdma.cpp -lhsa-runtime64 -o hsa_sdma && ./hsa_sdma
Agents found: cpu, gfx1034, 
Data for CPU to gfx1034
GPU to CPU: 00000000000000000000000000000011
CPU to GPU: 00000000000000000000000000000011
Count after op: 1
```

**Comment by cjatin** [comment]
I think this is the issue.

The device has 1 SDMA engine per: `/sys/devices/virtual/kfd/kfd/topology/nodes/1/properties:num_sdma_engines 1`

But rocr is reporting 2 via its mask: `00000000000000000000000000000011`

This causes HIP to use 2nd engine, when first one is busy with 1st memcpyasync and it fails:
```
:4:rocblit.cpp              :707 : 1316416892035 us: 3327951: [tid:0x7f4a18571100] HSA Async Copy on copy_engine=2 ...
:1:rocblit.cpp              :736 : 1316416892036 us: 3327951: [tid:0x7f4a18571100] HSA copy from host to device failed with code 4097
:1:rocvirtual.cpp           :1856: 1316416892037 us: 3327951: [tid:0x7f4a18571100] submitCopyMemory failed!
```

Will pass it on to rocr team. Thanks for reporting this.

**Comment by al42and** [comment]
Fixed in 5.7.0, thanks!

```console
$ hipcc hsa_sdma.cpp -lhsa-runtime64 -o hsa_sdma && ./hsa_sdma
Agents found: cpu, gfx1034, 
Data for CPU to gfx1034
Count after op: 0
GPU to CPU: 00000000000000000000000000000001
CPU to GPU: 00000000000000000000000000000001
Count after op: 1
```

**Comment by al42and** [comment]
> Interesting, so the second memcpy fails for some reason (presumably, HSA_STATUS_ERROR_INVALID_ARGUMENT), but the error is not propagated 😒.

That said, it would have been better if the HSA errors were properly reported to the user, not quietly ignored. A command failing for no reason is bad, but a command failing _silently_ for no reason is worse.
